### ISSUE TITLE
Unresolved conkeyrefs.

### ISSUE THREAD

**Issue by rui-doc-oliveira** [dita-ot/dita-ot#2687]
We're currently working on a ditamap where we have two topics where we are trying to use conkeyrefs. Both files have conkeyrefs in the title and in tags inside the body, such as a "p" and a "section".

I worked through the pipeline backwards and found by the time we get to the mappullImpl.xsl (when we first get an error on the output) the file is already missing all tags with conkeyrefs. The entire tag, and the content, are gone. Since the topic titles have conkeyrefs, we get errors in the ToC entries.

We've updated to the latest dita-ot we could find (DITA-OT version = 2.4.6+b074a2c) but still can't get conkeyref to work. We're using OxygenXML 19.0 and the editor resolves all conkeyrefs perfectly (in Author mode the editor displays the correct fetched text) and gives no errors or warnings.

I tried to look through dita-ot for references to @conkeyref that might help figure out where I need to debug and figure out what is wrong. Can't find anything. I find a lot of stuff for keyrefs, but nothing for conkeyrefs. Can't figure out where they're processed.

So basically... as anyone had a similar problem? Or can someone point me to where I should start looking, please?

**Comment by raducoravu** [comment]
Could you prepare a small DITA project (map + one or two topics) which could be used to reproduce the reported problem?

**Comment by rui-doc-oliveira** [comment]
Hi,

Thank you for the quick reply and I'm sorry for spamming. I had no intention of doing it. Sorry.

I'm having trouble adding the ZIP files to this post (the form tells me it doesn't support that file type although it's a simple ZIP file generated with WinRAR. I copied the ZIP files to my Dropbox, and hopefully you can access it from the links.

I made two versions. One using our folder structure and content, the other using a very simplified version of the content, and somewhat simplified folder structure. Both fail to include the conkeyref'ed content in the output. The ZIPs also include the log.txt, the resulting PDF and the full content of the temporary folder (I ran the dita.bat with -Dclean.temp=no)

Thank you so much for the quick reply and the support.

[Simplified ditamap, with our folder structure](https://www.dropbox.com/s/zxmdbetcla4gnt9/dita-test-as-is.zip?dl=0)
[Extremelly simplified ditamap, with reduced folder structure](https://www.dropbox.com/s/h4mrpg51mzncncm/dita-test-simplified.zip?dl=0)

**Comment by raducoravu** [comment]
I took a look at the extremely simplified example. The "topic.xml" was not referenced in the DITA Map.
I referenced it in the DITA Map and then published the DITA Map to PDF. The obtained PDF seems to properly expand conrefs:
[library-test.pdf](https://github.com/dita-ot/dita-ot/files/1006682/library-test.pdf)

Are you trying to generate the PDF directly from the DITA topic and not from a DITA Map referencing the topic?
If so, this has never worked as when publishing directly the topic the publishing engine does not know how to resolve keyrefs and conkeyrefs:
https://github.com/dita-ot/dita-ot/issues/1697

**Comment by rui-doc-oliveira** [comment]
Hi,

First of all thank you for the help and sorry for the delay response, I didn't see the email warning me about the result.

I understand our file structure is a bit strange, so I made the attached document. The document as a  diagram to explain the four files we use and screenshots of what we see in Oxygen XML (and works perfectly) and what is missing from the dita-ot generated PDF. 

I hope it helps, please let me know if I can provide more information.

[github.docx](https://github.com/dita-ot/dita-ot/files/1019037/github.docx)

**Comment by raducoravu** [comment]
Oh, I understand now. The DITA OT publishing engine has problems when publishing DITA projects having DITA content in folders outside of the folder where the main published DITA Map is located.
There is a "generate.copy.outer" parameter which can be set to "3" and which sometimes helps.
Also in our user's manual we give 2 more possible workarounds:
https://www.oxygenxml.com/doc/versions/19.0/ug-editor/topics/dita-ot-external-refs.html

**Comment by rui-doc-oliveira** [comment]
Thank you for the quick response and help.

I think We solved it. With the suggestion from the article you pointed out, thank you. Kind of. Adapted our PDF generator app to create a temporary ditamap at the highest folder in the folder structure, referencing the ditamap we want to generate. And once we have the generated PDF the app deletes the temporary ditamap.

At first glance, it works OK. 

We'll start testing it today, hopefully it works for the full ditamap repository. I'll report the outcome in a couple of days.

We've been using the generate.copy.outer at "3" for a few weeks, to solve keyref issues with the same folder structure. Had to change the front-matter.xsl to adapt to using cover images in a different folder for example.

I hope that in a future version of dita-ot we have a more stable handling of the situation, for generate.copy.outer at "3". If it's allowed, then it should work.

Your solution also helped solve an even more mysterious bug: had a ditamap that failed to generate correctly - keyref path issues - yet it would work perfectly if I just changed the ditamap name (and ditaval). Just the filenames. From "dk01-008-maintenance-1" to "dk01-008-maintenance-2". Spend days trying to debug this and the best I could come up with was each generation had different .list files in the fullditamap.list and fullditamapandtopic.list files. It's insane.

Again, thank you for your support!

**Comment by raducoravu** [comment]
About this problem:

> Your solution also helped solve an even more mysterious bug: had a ditamap that failed to generate correctly - keyref path issues - yet it would work perfectly if I just changed the ditamap name (and ditaval). Just the filenames. From "dk01-008-maintenance-1" to "dk01-008-maintenance-2"

I'm not sure what the problem was, maybe you reused the temporary and output folders between reruns.

**Comment by rui-doc-oliveira** [comment]
I rebooted the PC, cleaned up the Windows user Temp files and the dita output folder... am I missing any other cache, folder, or the like?

Anyway, the issue I initially reported is apparently totally fixed now. Thank you for the suggestion of the workaround. I'll close the issue.

**Comment by raducoravu** [comment]
👍
